### PR TITLE
make AsTable more careful when accepting argument

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/3062

### CODE REVIEW

File: src/other/utils.jl [hunk 1]
```diff
@@ -54,6 +54,22 @@ julia> select(df1, AsTable([:a, :b]) => ByRow(nt -> map(x -> x^2, nt)) => AsTabl
...
+                          AbstractVector{<:Integer},
+                          AbstractVector{Symbol},
+                          AbstractVector{<:AbstractString},
+                          Colon, Not, Between, All, Cols, Regex})
+            if !(cols isa AbstractVector &&
+                 all(col -> col isa Union{Integer, Symbol, AbstractString}, cols))
+                throw(ArgumentError("Unrecognized column selector $cols"))
+            end
+        end
+        return new(cols)
```
Comment (nalimilan) [inline src/other/utils.jl:71]: I find this a bit simpler to grasp:
```suggestion
        if cols isa Union{AbstractString, Symbol, Signed, Unsigned,
                          AbstractVector{<:Integer},
                          AbstractVector{Symbol},
                          AbstractVector{<:AbstractString},
                          Colon, Not, Between, All, Cols, Regex}) ||
            (cols isa AbstractVector &&
             all(col -> col isa Union{Integer, Symbol, AbstractString}, cols))
                return new(cols)
            else
                throw(ArgumentError("Unrecognized column selector $cols"))
            end
        end
```

File: test/index.jl [hunk 2]
```diff
@@ -529,4 +529,41 @@ end
...
+    @test_throws ArgumentError AsTable(Any[true, missing])
+    @test_throws ArgumentError AsTable([:A, missing])
+    @test_throws ArgumentError AsTable(Any[:A, missing])
+    @test_throws ArgumentError AsTable(1.0:1.0)
+    @test_throws ArgumentError AsTable([1.0])
+    @test_throws ArgumentError AsTable(Any[1.0])
+    @test_throws ArgumentError AsTable([:a, "a", 1, true, 1.0])
+    @test_throws ArgumentError AsTable(:a => :b)
+
+    @test_nowarn AsTable(Union{Bool, Missing}[true, false])
```
Comment (nalimilan) [inline test/index.jl:548]: What kind of warning could this print?

File: src/abstractdataframe/selection.jl [hunk 3]
```diff
@@ -343,7 +343,7 @@ normalize_selection(idx::AbstractIndex, @nospecialize(sel), renamecols::Bool) =
         idx[sel]
     catch e
         if e isa MethodError && e.f === getindex && e.args === (idx, sel)
-            throw(ArgumentError("Unrecognized column selector: $sel"))
+            throw(ArgumentError("Unrecognized column selector: $sel in AsTable constructor"))
```
Comment (nalimilan) [inline src/abstractdataframe/selection.jl:346]:
```suggestion
            throw(ArgumentError("Unrecognized column selector $sel in AsTable constructor"))
```

File: test/index.jl (hunk 2 above)
Comment (bkamins) [reply test/index.jl:548]: The name of the macro is misleading. What it does is more general:

> Test whether evaluating expr results in empty stderr output

so essentially I test if the expression was executed cleanly. I could drop `@test_nowarn`, but it does not harm to have it I think.

File: src/other/utils.jl (hunk 1 above)
Comment (bkamins) [reply src/other/utils.jl:71]: OK - I thought you would find it less readable :).

File: test/index.jl (hunk 2 above)
Comment (nalimilan) [reply test/index.jl:548]: I'd be inclined to drop these as reading them you'd think a warning could be thrown somewhere, but as you like. Could also test that the `cols` field contains what we expect it to, though it's kind of obvious.

Comment (bkamins) [reply test/index.jl:548]: OK - I will change this to the test you propose.

File: test/index.jl [hunk 4]
```diff
@@ -545,25 +545,24 @@ end
...
-    @test_nowarn AsTable(0x1)
-    @test_nowarn AsTable([:a, :b])
-    @test_nowarn AsTable(["a", "b"])
-    @test_nowarn AsTable([1, 2])
-    @test_nowarn AsTable(Integer[1, true])
+    @test AsTable(Union{Bool, Missing}[true, false]).cols == Union{Bool, Missing}[true, false]
+    @test AsTable([:a, "a", 1, true]).cols == [:a, "a", 1, true]
+    @test AsTable([]).cols == []
+    @test AsTable(Int[]).cols == Int[]
+    @test AsTable(Symbol[]).cols == Symbol[]
```
Comment (nalimilan) [inline test/index.jl:552]:
```suggestion
    @test AsTable(Int[]).cols == []
    @test AsTable(Symbol[]).cols == []
```

File: test/index.jl [hunk 5]
```diff
@@ -545,25 +545,24 @@ end
...
-    @test_nowarn AsTable(All())
-    @test_nowarn AsTable(r"x")
-    @test_nowarn AsTable("a")
-    @test_nowarn AsTable(1)
-    @test_nowarn AsTable(0x1)
-    @test_nowarn AsTable([:a, :b])
-    @test_nowarn AsTable(["a", "b"])
-    @test_nowarn AsTable([1, 2])
-    @test_nowarn AsTable(Integer[1, true])
+    @test AsTable(Union{Bool, Missing}[true, false]).cols == Union{Bool, Missing}[true, false]
```
Comment (nalimilan) [inline test/index.jl:548]:
```suggestion
    @test AsTable(Union{Bool, Missing}[true, false]).cols == [true, false]
```

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**